### ISSUE TITLE
race condition in 'archive'

### ISSUE THREAD

**Issue by justinfisk** [Applied-GeoSolutions/gips#17]
I believe I have encountered a race condition in 'archive'. The symptom is that gips_process will die with the following error: 

```
  File "/home1/jfisk/venv/lib/python2.7/site-packages/gips-0.8.1-py2.7.egg/gips/data/core.py", line 
464, in _archivefile
    raise Exception('Problem adding %s to archive: %s' % (filename, e))
Exception: Problem adding /nobackup/jfisk/data/landsat/stage/LC80270272014198LGN00.tar.gz to archive
: [Errno 17] File exists

Data processing error: Problem adding /nobbackup/jfisk/data/landsat/stage/LC80270272014198LGN00.tar.g
z to archive: [Errno 17] File exists
```

I believe it to be related to multiple processes fetching landsat scenes simultaneously, and the 'archive' step moving files that don't belong to it.

**Comment by bhbraswell** [comment]
It's also possible that archive will move partially downloaded files from other processes.

Rob

**Comment by ircwaves** [comment]
Doesn't smell to racy to me ;).  But from what you've said here, there certainly is not any concurrency safety in the fetching and archiving.  Are you in a situation where you know you have two processes that are both downloading and archiving a file? 

Should archiving have some amount of exclusivity?  My gut says no.  Perhaps there should just be a check of the valididty of the asset when archiving encounters an errno-17.  If the asset is valid and of the same name, ditch the downloaded file and move on.  If it is corrupt, verify** the newly downloaded asset, and replace?

++ Ian (thinking out loud after days of high fever -- so I could be whacked)

** `verify()` seems like maybe something that should be added to the asset class

**Comment by justinfisk** [comment]
Yes, I should have been more specific...

I started with an empty data directory, and was running multiple gips_process commands (with --fetch). Each gips_process command was operating on a specific tile, so in no case should they have fetched the same tile. I have reproduced the same conditions and same error.  

I am not convinced that verify() would solve the problem in the case of another process moving a partially downloaded file.

**Comment by bhbraswell** [comment]
But also I think that individual fetch/ archive processes should be isolated so that they don't try to sweep out the common stage dir while in the middle of downloads

**Comment by ircwaves** [comment]
`verify` certainly wouldn't address the problem.  That was an additional idea to mitigate the impact of the collision in a way that left the repository as clean as possible.

But, you say that each `gips_process` is working on a separate tile.  That really doesn't smell racy.  If one is working on `027027` and another on `027028`, then they should certainly never step on each other.

**Comment by ircwaves** [comment]
Oh, that comment was actually made before Rob's, but I didn't hit enter.

So, landsat fetch needs to use the `tempfile`module so that each process has their own sandbox.  That should be hard to fix.

**Comment by demiurg** [comment]
Is there any way to replicate reliably the issue?

**Comment by justinfisk** [comment]
As with most things that depend on the timing of competing processes, it is not likely reliably replicated. (Although it is not rare in my case of running many processes simultaneously.) 

But it is easy to see where it would happen. In the 'archive' method, it gets a list of files based on the asset 'pattern'. That pattern grabs files which are not necessarily ready or related to the tile that should be being processed. Later, the _archiveFile method actually moves a file from that matched list. Any other process which generated its list based on 'pattern' before the files are moved would have a list that contains files that could have already moved by the time it gets to is own _archiveFile

Process A | Process B
-------------- | -------------
arrives at archive step and grabs list of files in stage that match asset 'pattern' | 
 |  arrives at archive step and grabs list of files in stage that match asset 'pattern'
moves files into place in _archveFile |
 | can't move files into place

**Comment by ircwaves** [comment]
The issue is inherent.  The landsat fetch uses a common staging directory, and it needs to use tempdirs inside of that staging directory.

**Comment by demiurg** [comment]
Doesn't 'unique tempdir' just delay the potential conflict?

**Comment by justinfisk** [comment]
Not sure I see how. Seems to me like tempdirs would solve. If the 'archive' method gets a list of files that only owns, then no other process could move those files into place. I believe tempdirs automagically cleans up after itself, so nothing would be left behind if something else went wrong. So what additional conflict am I missing?

**Comment by ircwaves** [comment]
I suppose the remaining conflict is that B cannot move its copy of asset X into the archive because A has already moved its copy of X into the archive.  In which case, B's copy of X is redundant and can be dropped.

**Comment by bhbraswell** [comment]
Why are A and B getting the same asset? seems like poor planning.

Rob

**Comment by justinfisk** [comment]
In my case, they weren't.

**Comment by ircwaves** [comment]
In that case, there would be no conflict.

**Comment by bhbraswell** [comment]
There would, because B will call archive before it does anything, and thus try to sweep the data from stage (including partially downloaded files) into the archive. I think this would be solved by dedicated temp dirs.

**Comment by justinfisk** [comment]
Yes, I think tempdirs solves the first problem (I think we are on the same page about that, except perhaps that one Russian guy.) 

But it might be reasonable to assume that we, as a group, are not capable of 'planning' well enough to prevent two people from ever requesting the same asset at the same time.

**Comment by demiurg** [comment]
Poor or non existent planning is a typical use case for most software. This is quite likely, I hope it can be fixed with an ACID database, which sqlite isn't.

**Comment by bhbraswell** [comment]
I think that the new revamped inventory mechanism will allow for locking one particular operation like a specific download, that might conflict with another one.

**Comment by demiurg** [comment]
Yeah, it should, I just don't know if it can with sqlite in particular.

**Comment by ircwaves** [comment]
Rob -- 

In suggesting the use of temp dirs, I omitted the idea that the archive call (subsequent to a fetch into a tempdir) should be passed that temp directory's path.  That way it will only sweep up that which was downloaded into that directory.  

But I guess in the loosely regulated form it currently exists, stage can always get swept out from under you.

**Comment by ircwaves** [comment]
related to issue #115

**Comment by ircwaves** [comment]
In addition to implementing #115, perhaps `gips_archive` should also lock the directory which it is working on?  115 will make this mostly useless, but it may save headaches in other places.  And may be sufficient for CLI level concurrency safety.
